System/OsString/Internal/Types.hs

```diff
@@ -0,0 +1,233 @@
+-- dealing with the internals isn't generally recommended, but supported
+-- in case you need to write platform specific code.
+newtype OsString = OsString PlatformString
+  deriving (Typeable, Generic, NFData)
+
+-- | Byte equality of the internal representation.
+instance Eq OsString where
+  (OsString a) == (OsString b) = a == b
+
+-- | Byte ordering of the internal representation.
+instance Ord OsString where
+  compare (OsString a) (OsString b) = compare a b
+
+-- | Encodes as UTF16 on windows and UTF8 on unix.
+instance IsString OsString where 
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+    fromString = OsString . WS . encodeUtf16LE
+#else
+    fromString = OsString . PS . encodeUtf8
+#endif
+
+
+-- | \"String-Concatenation\" for 'OsString. This is __not__ the same
+-- as '(</>)'.
+instance Monoid OsString where 
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+    mempty      = OsString (WS BS.empty)
+#if MIN_VERSION_base(4,16,0)
+    mappend = (<>)
+#else
+    mappend (OsString (WS a)) (OsString (WS b))
+      = OsString (WS (mappend a b))
+#endif
+#else
+    mempty      = OsString (PS BS.empty)
+#if MIN_VERSION_base(4,16,0)
+    mappend = (<>)
+#else
+    mappend (OsString (PS a)) (OsString (PS b))
+      = OsString (PS (mappend a b))
+#endif
+#endif
+#if MIN_VERSION_base(4,11,0)
+instance Semigroup OsString where 
+#if MIN_VERSION_base(4,16,0)
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+    (<>) (OsString (WS a)) (OsString (WS b))
+      = OsString (WS (mappend a b))
+#else
+    (<>) (OsString (PS a)) (OsString (PS b))
+      = OsString (PS (mappend a b))
+#endif
+#else
+    (<>) = mappend
+#endif
+#endif
+
+
+instance Lift OsString where
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+  lift (OsString (WS bs))
+    = [| OsString (WS (BS.pack $(lift $ BS.unpack bs))) :: OsString |]
+#else
+  lift (OsString (PS bs))
+    = [| OsString (PS (BS.pack $(lift $ BS.unpack bs))) :: OsString |]
+#endif
+#if MIN_VERSION_template_haskell(2,17,0)
+  liftTyped = TH.unsafeCodeCoerce . TH.lift
+#elif MIN_VERSION_template_haskell(2,16,0)
+  liftTyped = TH.unsafeTExpCoerce . TH.lift
+#endif
+
+-- | Decodes as UTF-16 on windows.
+--
+-- Decodes as UTF-8 on unix and replaces invalid chars with unicode replacement
+-- char U+FFFD.
+instance Show OsString where
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+  show (OsString (WS bs)) = ('\"': decodeUtf16LEWith lenientDecode bs) <> "\""
+#else
+  show (OsString (PS bs)) = ('\"': decodeUtf8With lenientDecode bs) <> "\""
+#endif
+
+-- | Encodes as UTF-8 on unix and UTF-16LE on windows.
+instance Read OsString where
+#if defined(mingw32_HOST_OS) || defined(__MINGW32__)
+  readsPrec p str = [ (OsString $ WS $ encodeUtf16LE x, y) | (x, y) <- readsPrec p str ]
+#else
+  readsPrec p str = [ (OsString $ PS $ encodeUtf8 x, y) | (x, y) <- readsPrec p str ]
+#endif
+
+
+-- | Newtype representing a limited Codepoint.
```

Comment:
"Limited codepoints" are called code units, and it is a better word than `Char`.